Bug 1857246: Retrying in case of conflict during Config update

ricardomaraschini opened the pull request:
We have multiple controllers concurrently writting the status of the
config, this commit introduces retries in case of any of them updated
the config resource before us. This is clearly a workaround, as the
main controller may create storage if we fail to update the
configuration here it may ended up creating storage again on the next
cycle, hence the need to make sure we succeed.

ricardomaraschini commented:
/assign @dmage

ricardomaraschini commented:
/retest

dmage commented:
@ricardomaraschini can you create an e2e test that checks that after the bootstrap storage is created only once?

ricardomaraschini commented:
/retest

ricardomaraschini commented:
> @ricardomaraschini can you create an e2e test that checks that after the bootstrap storage is created only once?

As to have tests for every platform would be overkill I changed the error message and implemented a test to make sure it is not happening. PTAL.

dmage commented on test/e2e/bootstrap_test.go:
```diff
@@ -0,0 +1,33 @@
+)
+
+func TestBootstrapFailToUpdateSpec(t *testing.T) {
+	te := framework.Setup(t)
```
Use SetupAvailableImageRegistry instead of this bunch of Deploy/Wait/Ensure calls.

dmage commented:
/lgtm

dmage commented:
/hold
it doesn't work

dmage commented on pkg/operator/controller.go:
```diff
@@ -247,14 +248,26 @@ func (c *Controller) sync() error {
+			updatedCR, err = c.listers.RegistryConfigs.Get(defaults.ImageRegistryResourceName)
+			if err != nil {
+				return err
 			}
```
the object from the lister shouldn't be mutated, DeepCopy may help

ricardomaraschini replied:
done.

dmage commented:
/retest

dmage commented:
/lgtm

dmage commented:
/hold cancel

dmage commented:
/cherrypick release-4.5
